README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

Overview To start using the installed package:

``` r
library(radsafer)
```

The radsafer package was developed with these goals:

- Provide functions that are commonly used in radiation safety

- Provide easy access to data provided in the RadData package

- Share some less commonly-used functions that may be of significant
  value to workers in instrumentation and modeling

Related functions are identified as members of these families:

- decay corrections

- radionuclides

- rad measurements

- mcnp tools

A few functions stand alone.

This framework makes it a little easier to find the function you want -
the help file for each function lists related functions.

### Decay Correction Functions

Radsafer includes several functions to manage radioactive decay
corrections:

**dk_correct** provides decay corrected activity or activity-dependent
value, such as instrument response rate or dose rate. The computation is
made either based on a single radionuclide, or based on user-provided
half-life, with time unit. The differential time is either computed
based on dates entered or time lapsed based on the time unit.

Obtain a correction factor for a source to be used today based on a
calibration on `date1`. Allow `date2` to be the default system date.

``` r
dk_correct(half_life = 10,
           time_unit = "y",
           date1 = "2010-01-01")
#>  half_life RefValue    RefDate   TargDate  dk_value
#>         10        1 2010-01-01 2023-07-11 0.3916907
```

Use this function to correct for the value needed on dates it was used.
Let the function obtain the half-life from the ICRP Publication 107
decay data in the `RadData` R package. In the example, we have a disk
source with an original count rate of 10000 cpm:

``` r
dk_correct(RN_select = "Sr-90",
           date1 = "2005-01-01",
           date2 = c("2009-01-01","2009-10-01"),
           A1 = 10000)
#>     RN half_life units decay_mode
#>  Sr-90     28.79     y         B-
#> 
#>     RN RefValue    RefDate   TargDate dk_value
#>  Sr-90    10000 2005-01-01 2009-01-01 9081.895
#>  Sr-90    10000 2005-01-01 2009-10-01 8919.954
```

Reverse decay - find out what readings should have been in the past
given today’s reading of 3000

``` r
dk_correct(RN_select = "Cs-137", 
  date1 = "2019-01-01", 
  date2 = c("2009-01-01","1999-01-01"), 
  A1 = 3000)
#>      RN half_life units decay_mode
#>  Cs-137   30.1671     y         B-
#> 
#>      RN RefValue    RefDate   TargDate dk_value
#>  Cs-137     3000 2019-01-01 2009-01-01 3774.795
#>  Cs-137     3000 2019-01-01 1999-01-01 4749.991
```

Other decay functions answer the following questions:

- How long does it take to decay something with a given activity, or how
  old is a sample if it has decayed from? **dk_time**

- Given a percentage reduction in activity, how many half-lives have
  passed.**dk_pct_to_num_half_life**

- Given two or more data points, estimate the half-life:
  **half_life_2pt**

### radionuclides

Search by alpha, beta, photon or use the general screen option.

`RN_search_phot_by_E` allows screening based on energy, half-life, and
minimum probability. Also available are `RN_search_alpha_by_E`,
`RN_search_beta_by_E`, and `bin_screen_phot`. `RN_bin_screen_phot`
allows limiting searches to radionuclides with emissions in an energy
bin of interest with additional filters for not having photons in other
specified energy bins. Results for all these search functions may be
plotted with `RN_plot_spectrum`.

Here’s a search for photon energy between 0.99 and 1.01 MeV, half-life
between 13 and 15 minutes, and probability at least 1e-3

``` r
search_results <- RN_search_phot_by_E(0.99, 1.01, 13 * 60, 15 * 60, 1e-3)
```

| RN     | code_AN |   E_MeV |      prob | half_life | units | decay_constant |
|:-------|:--------|--------:|----------:|----------:|:------|---------------:|
| Pr-136 | G       | 0.99100 | 0.0016768 |     13.10 | m     |      0.0008819 |
| Pr-136 | G       | 1.00070 | 0.0503040 |     13.10 | m     |      0.0008819 |
| Re-178 | G       | 1.00440 | 0.0057600 |     13.20 | m     |      0.0008752 |
| Pr-147 | G       | 0.99597 | 0.0083220 |     13.40 | m     |      0.0008621 |
| Nb-88  | G       | 0.99760 | 0.0041000 |     14.50 | m     |      0.0007967 |
| Mo-101 | G       | 1.00740 | 0.0017300 |     14.61 | m     |      0.0007907 |
| Sm-140 | G       | 0.99990 | 0.0012000 |     14.82 | m     |      0.0007795 |

``` r
RN_plot_spectrum(search_results)
#> [1] "No matches"
```

You can also plot a spectrum in one step, skipping the data save:

``` r
 RN_plot_spectrum(
   desired_RN = c("Pu-238", "Pu-239", "Am-241"), rad_type = "A",
   photon = FALSE, prob_cut = 0.01, log_plot = 0)
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />

The `RN_index_screen` function helps find a radionuclide of interest
based on decay mode, half-life, and total emission energy.

In this example, we search for radionuclides decaying by spontaneous
fission with half-lives between 6 months and 2 years.

``` r
RNs_selected <- RN_index_screen(dk_mode = "SF", min_half_life_seconds = 0.5 * 3.153e7, max_half_life_seconds = 2 * 3.153e7)
```

| RN     | half_life | units |
|:-------|----------:|:------|
| Es-254 |     275.7 | d     |
| Cf-248 |     334.0 | d     |

Other radionuclides family functions:

- Obtain a specific activity with **RN_Spec_Act**.

- Find a potential radionuclide parent with **RN_find_parent**.

``` r

RN_find_parent("Th-230")
#>       RN
#> 1 Ac-230
#> 2 Pa-230
#> 3  U-234
```

### rad measurements functions

**air_dens_cf** Correct *vented ion chamber readings* based on
difference in air pressure (readings in degrees Celsius and mm Hg):

``` r
air_dens_cf(T.actual = 30, P.actual = 760, T.ref = 20, P.ref = 760)
#> [1] 1.034112
```

Let’s try it out combined with the instrument reading:

``` r
rdg <- 100
(rdg_corrected <- rdg * air_dens_cf(T.actual = 30, P.actual = 760, T.ref = 20, P.ref = 760))
#> [1] 103.4112
```

**neutron_geom_cf**

Correct for *geometry* when reading a close *neutron* source. Example:
neutron rem detector with a radius of 11 cm and source near surface:

``` r
neutron_geom_cf(11.1, 11)
#> [1] 0.7236467
```

**disk_to_disk_solid_angle**

Correct for a mismatch between the *source calibration* of a *counting
system* and the item being measured. A significant factor in the
counting efficiency is the solid angle from the source to the detector.
You can also check for the impact of an item not being centered with the
detector.

Example: You are counting an air sample with an active collection
diameter of 45 mm, your detector has a radius of 25 mm and there is a
gap between the two of 5 mm. (The function is based on radius, not
diameter so be sure to divide the diameter by two.) The relative solid
angle is:

``` r
(as_rel_solid_angle <- as.numeric(disk_to_disk_solid_angle(r.source = 45/2, gap = 20, r.detector = 12.5, runs = 1e4, plot.opt = "n")))
#> [1] 0.047469927 0.002120278
```

An optional plot is available in 2D or 3D:

``` r
  library(ggplot2)
theme_update(# axis labels
             axis.title = element_text(size = 7),
             # tick labels
             axis.text = element_text(size = 5),
             # title 
             title = element_text(size = 5))
(as_rel_solid_angle <- as.numeric(disk_to_disk_solid_angle(r.source = 45/2, gap = 20, r.detector = 12.5, runs = 1e4, plot.opt = "3d")))
```

<img src="man/figures/README-unnamed-chunk-16-1.png" width="50%" style="display: block; margin: auto;" />

    #> [1] 0.047585442 0.002137657

Continuing the example: the only calibration source you had available
with the appropriate isotope has an active diameter of 20 mm. Is this a
big deal? Let’s estimate the relative solid angle of the calibration,
then take a ratio of the two.

``` r
(cal_rel_solid_angle <- disk_to_disk_solid_angle(r.source = 20, gap = 20, r.detector = 12.5, runs = 1e4, plot.opt = "n"))
#>   mean_eff         SEM
#>  0.0528169 0.002243531
```

Correct for the mismatch:

``` r
(cf <- cal_rel_solid_angle / as_rel_solid_angle)
#>  mean_eff      SEM
#>  1.109938 1.049528
```

This makes sense - the air sample has particles originating outside the
source radius, so more of them will be lost, thus an adjustment is
needed for the activity measurement.

**scaler_sim**

*Scaler counts*: obtain quick distributions for parameters of interest:

``` r
scaler_sim(true_bkg = 50, true_samp = 10, ct_time = 20, trials = 1e5)
```

<img src="man/figures/README-unnamed-chunk-19-1.png" width="50%" style="display: block; margin: auto;" />

**rate_meter_sim**

*Rate meters*: In the ratemeter simulation, readings are plotted once
per second for a default time of 600 seconds. The meter starts with a
reading of zero and builds up based on the time constant. Resolution
uncertainty is established to express the uncertainty from reading an
analog scale, including the instability of its readings. Many standard
references identify the precision or resolution uncertainty of analog
readings as half of the smallest increment. This should be considered
the single coverage uncertainty for a very stable reading. When a
reading is not very stable, evaluation of the reading fluctuation is
evaluated in terms of numbers of scale increments covered by meter
indication over a reasonable evaluation period. Example with default
time constant:

``` r
rate_meter_sim(cpm_equilibrium = 270, meter_scale_increments = seq(100, 1000, 20))
```

<img src="man/figures/README-unnamed-chunk-20-1.png" width="50%" style="display: block; margin: auto;" />

To estimate *time constant*, use `tau.estimate`

### Some stand-alone functions

#### Stay-time computation

Given a dose rate, dose allowed, and a safety margin (default = 20%),
calculate stay time with: `stay_time`

``` r
stay_time(dose_rate = 120, dose_allowed = 100, margin =  20)
#> [1] "Time allowed is 40 minutes"
#> [1] 40
```

#### half-value and tenth value computation

Given a set of radiation levels through varying thicknesses of material,
compute the half-values, tenth-values, and homogeneity coefficients.

``` r
hvl(x = mm_Al, y = mR_h) 
#>   thickness response        mu      hvl      tvl        hc
#> 1         0    7.428        NA       NA       NA        NA
#> 2         1    6.272 0.1691614 4.097550 13.61177 0.9676121
#> 3         2    5.325 0.1636826 4.234704 14.06738 0.9810918
#> 4         3    4.535 0.1605876 4.316317 14.33850 0.9745802
#> 5         4    3.878 0.1565055 4.428899 14.71248 0.9984235
#> 6         5    3.317 0.1562588 4.435892 14.73571        NA
```

### mcnp tools functions

If you create MCNP inputs, these functions may be helpful:

**mcnp_sdef_erg_line** Obtain emission data from the `RadData` package
for use with the radiation transport code, MCNP, when you want a “line”
source from individual radionuclides. Use **mcnp_sdef_erg_hist** when
you have histogram data and want to format it nicely for MCNP.

**mcnp_matrix_rotations**

- Determine the entries needed for MCNP *coordinate transformation
  rotation*

``` r
mcnp_matrix_rotations("z", 90)
#> [1]  0  1  0 -1  0  0  0  0  1
```

**mcnp_cone_angle** This is just the square of the tangent of the angle
in radians, but the argument used here is degree.

``` r
mcnp_cone_angle(30)
#> [1] 0.3333333
```

**mcnp_mesh_bins** This helps identify the x, y, or z parameter of a
simple geometric mesh tally. Select a center of interest, the width of
this center mesh, then minimum and maximum limits on the extent of the
entire mesh. Repeat for the other dimensions.

``` r
mcnp_mesh_bins(target = 30, width = 10, lowest_less = 0, highest_less = 15, highest_high = 304.8, lowest_high = 250)
#>  low_set high_set width numblocks
#>        5      295    10        29
```

**mcnp_plot_out_spec**

For *MCNP outputs*, plot the results of a tally with *energy bins*. The
fastest way to do this is with `mcnp_scan2plot`.

Alternatively, if you want to get your data into R, first save it in a
text file and import it to R. (Base R provides methods with `read.table`
or you might prefer options from the `readr` or `readxl` packages.) Or
you can copy and paste your data using `mcnp_scan_save`. You can then
plot with `mcnp_plot_out_spec` (below) or design your own plot.

``` r
mcnp_plot_out_spec(photons_cs137_hist, 'example Cs-137 well irradiator')
```

<img src="man/figures/README-unnamed-chunk-28-1.png" width="75%" style="display: block; margin: auto;" />


NEWS.md
# radsafer (version 2.3.0)
***2023-07-11*** Adding new function to produce geometry parameters for MCNP geometric mesh tallies. Correcting documentation for `RN_plot_spectrum`. Adding argument check for this and `RN_save_spectrum` - the photon argument must be logical.

# radsafer (version 2.2.8)
***2023-04-27*** Adding new function to compute half-value layers and tenth-value layers. Combining and renaming functions to make energy cards for MCNP. Tweaks to many functions to enhance use of numeric results. 

# radsafer (version 2.2.6)
***2022-02-02*** Two functions for providing source energy distributions for MCNP have been updated to make them easier to use. Previously, version 2.2.2 implemented a CRAN requirement in `mcnp_si_sp_hist` and `mcnp_si_sp_RD` to ensure users gave consent to any file writing. This was implemented with a required 'y' response to a dialog in the console. This has proven to be cumbersome and prevented use of these functions in R markdown files. This version changes the permission to a function parameter that defaults to 'n', retaining the permission requirement while allowing greater ease of use.

# radsafer (version 2.2.5)
***2021-04-12*** Updated branch_4 title to match repaired RadData index which previously dropped the 4. 

# radsafer (version 2.2.4)
***2020-12-31*** Corrected the column assignment in the four-column MCNP source copy and paste functions in mcnp_scan_save and mcnp_scan2plot. (Prior version incorrectly used bin number for energy.)

# radsafer (version 2.2.3)
***2020-09-09*** Add option to scan biased histogram (mcnp_scan_save). Remove duplicate to this function (mcnp_scan2spec). Return data frame from RN_info. Fix functions that have return without () per CRAN directive.

# radsafer (version 2.2.2)
***2020-07-23*** Small change to require user to allow file to be created or appended when creating si\_sp.txt files for MCNP input energy cards. This is required by CRAN.

# radsafer (version 2.2.1)

***2020-06-07*** Removed deprecated search functions that have been
replaced by searches starting with RN.

# radsafer (version 2.2.0)

***2020-02-21*** Improved plotting functions – discrete energy spectra
with lollipop plots (`RN_plot_spectrum`) and histograms plotted with
steps between energy bins `ggplot2::geom_step(direction = 'vh')` (in
`mcnp_plot_out_spec`). Added the `ggtheme` package’s `theme_calc` for
radionuclide plots (`RN_plot_spectrum`). In case users want to just get
a plot form MCNP or, possibly other histogram data, they can now do it
with one step (`mcnp_scan2plot`). The scan-in function, now called
`mcnp_scan_save`, can now scan in 4-column source histogram data from
output files in addition to the three-column output spectra. The two
column scan went away for lack of any known application, but can be put
back in if anyone needs it - use the maintainer email listed in the
package description. Some function renaming was done to make it easier
for users to select a helpful function, for example, `search_alpha_by_E`
is now `RN_search_alpha_by_E`. Renamed functions will have a warning to
use the new function name and these functions will be removed in a later
version. Previously deprecated functions were also removed and
documentation was improved.

# radsafer (version 2.1.0)

***2019-12-17*** Many improvements in rad measurements family plots. Bug
corrections in mcnp tools family, especially mcnp\_si\_sp\_hist, plus
split off of copy and paste version as mcnp\_si\_sp\_hist\_scan. New
function in radionuclides family - RN\_find\_parents - to find out where
a radionuclide may have decayed from. A vignette for overview of the
package is added.

# radsafer (version 2.0.1)

***2019-05-27*** This version is to be submitted to CRAN. Bug removal on
dk\_activity. Updated examples in many functions to identify parameters
at least once.

# radsafer (version 2.0.0.9000)

***2019-05-23*** Repaired a bug in dk\_activity. Working on improvements
in overall ease of use. Expecting an update to CRAN in a couple of
weeks.

# radsafer (version 2.0.0)

***2019-05-05*** Hurray\! This is the grand opening of radsafer\! The
ICRP 107 data in RadData is now effectively utilized for radionuclide
screening, plots, and use with MCNP. MCNP utilities have been expanded -
see the Readme file for details.

# radsafer (version 1.0.9000)

***2019-04-21***

New functions have been added to the development version, 1.0.0.9000.
Several of the existing functions have had slight changes, including
name changes that will help make the package easier to use going
forward.

# radsafer (version 1.0.0)

***2019-04-06***

Big news\! **radsafer** is now available from CRAN\! So, we should be
announcing the grand opening, right?

But wait. **RadData** (<https://github.com/markhogue/RadData>) has also
just been issued by CRAN. So, this is the soft opening for just a few
close friends because a much more compelling version of **radsafer**
will be out soon. It is going to have some really helpful functions for
accessing, screening, plotting, and making inputs from **RadData**.

Here’s the backstory of **radsafer** and **RadData**. Here at radsafer
world headquarters, we were all excited when we figured out what we
could do with access to the ICRP 107 Nuclear Decay Data for Dosimetric
Calculations. So excited, we wanted to share it, along with a bunch of
other generally helpful health physics functions. But then, we found we
couldn’t meet CRAN’s size limit of 5 MB with all that data. Then, we had
other issues to deal with and were not sure if **RadData** would ever
get off the ground. So, we pursued, in parallel, issuing a lighter
**radsafer** and **RadData**. And now, suddenly, success has arrived at
the same time on both projects\! We’re so psyched\!

So, stay tuned, friends. **radsafer** will be out with version 1.1.0

R/dk_correct.R
#' Correct activity-dependent value based on radioactive decay.
#'
#' @family decay corrections
#'
#' @description Decay-corrected values are provided. Either a single or multiple
#'   values are computed. The computation is made either based on a single
#'   radionuclide, or based on user-provided half-life, with time unit. The
#'   differential time is either computed based on dates entered or time lapsed
#'   based on the time unit. Time units must be consistent. Decay-correct a
#'   source to today's date by assigning a reference `date1` and allowing
#'   default `date2`, the system date.
#'
#' @param RN_select identify the radionuclide of interest in the format,
#'   *"Es-254m"* Required unless `half_life` is entered.
#' @param half_life Required if `RN_select` is not provided.
#' @param time_unit, acceptable values are *"years"*, *"days"*,
#'   *"hours"*, *"minutes"*, and *"seconds"*. May be shortened to
#'   *"y"*, *"d"*, *"h"*, *"m"*, and *"s"*. Required if
#'   `half_life` or `time_lapse` are to be entered.
#' @param time_lapse a single value or vector of values representing time lapsed
#'   since `date1`, with units identified in `time_unit`. Positive values
#'   represent time past `date1`. Negative values represent time before `date1`.
#'   Required unless `date1` is entered.
#' @param date1 Reference date - Required unless using `time_lapse`. Format is
#'   required to be date-only: *"YYYY-MM-DD"* (e.g. *"1999-12-31"*).
#'   If `half_life` is short relative to calendar dates, use `time_lapse`
#'   instead.
#' @param date2 Date or dates of interest. Default is today's date, obtained
#'   from the computer operating system.
#' @param A1 The reference activity or related parameter, such as count rate or
#'   dose rate. Default value is 1, resulting in a returned value that may be
#'   used as a correction factor.
#' @param num Set for TRUE to facilitate as.numeric results. Default = FALSE.
#' @return Decay adjusted activity or related parameter. See `A1`.
#' @examples
#'
#' # RN_select and date1 (saving numerical data)
#' my_dks <- dk_correct(
#'   RN_select = "Sr-90",
#'   date1 = "2009-01-01",
#'   date2 = "2019-01-01",
#'   num = TRUE
#' )
#'
#' #   RN_select and time_lapse (random sample)
#' dk_correct(
#'   RN_select = base::sample(RadData::ICRP_07.NDX$RN, 1),
#'   time_lapse = 1:10,
#'   time_unit = base::sample(c("y", "d", "h", "m", "s"), 1)
#' )
#'
#' #   half_life and date1
#' dk_correct(
#'   half_life = 10,
#'   time_unit = "y",
#'   date1 = "2009-01-01",
#'   date2 = c(
#'     "2015-01-01",
#'     "2016-01-01",
#'     "2017-01-01"
#'   )
#' )
#'
#' #   half_life and time_lapse
#' dk_correct(
#'   half_life = 10,
#'   time_lapse = 10,
#'   time_unit = "y"
#' )
#'
#' # decay to today
#' dk_correct(RN_select = "Sr-90", date1 = "2009-01-01")
#'
#' # reverse decay - find out what readings should have been in the past given today's reading of 3000
#' dk_correct(
#'   RN_select = "Sr-90",
#'   date1 = "2019-01-01",
#'   date2 = c("2009-01-01", "1999-01-01"),
#'   A1 = 3000
#' )
#' @export
dk_correct <- function(RN_select = NULL,
                       half_life = NULL,
                       time_unit = NULL,
                       time_lapse = NULL,
                       date1 = NULL,
                       date2 = Sys.Date(),
                       A1 = 1,
                       num = FALSE) {
  #
  # convert dates to date format, if needed
  if (!is.null(date1)) date1 <- as.Date(date1)
  if (!is.null(date2)) date2 <- as.Date(date2)

  # if time_lapse is entered as integer, convert
  if (is.integer(time_lapse)) time_lapse <- as.numeric(time_lapse)
  # arg checks~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  #
  #  Check for compatible argument set

  if (!is.null(RN_select) & !is.character(RN_select)) {
    stop("RN_select must be entered in quotes")
  }
  # Should have either RN_select or half_life

  if ((is.null(RN_select) + is.null(half_life)) != 1) {
    stop("Enter either RN_select or half_life, not both.")
  }

  # Should have either time_lapse or date1
  if ((is.null(time_lapse) + is.null(date1)) != 1) {
    stop("Enter either time_lapse or date1, not both.")
  }

  #  half_life check

  if (!is.null(half_life)) {
    if (!is.numeric(half_life)) {
      stop("half_life must be a number.")
    }
  }

  # time_unit check - time_unit needed for either time_lapse or half_life
  if (!is.null(c(time_lapse, half_life))) {
    if (is.null(time_unit)) stop("time_unit is needed to work with time_lapse and half_life.")
  }

  if (!is.null(time_unit)) {
    if (!time_unit %in% c(
      "years", "days", "hours", "minutes", "seconds",
      "y", "d", "h", "m", "s"
    )) {
      stop("time_unit is not one of the acceptable options:
       'years', 'days', 'hours','minutes', 'seconds',
        'y', 'd', 'h', 'm', 's' ")
    }
  }

  # check date formats
  if (!is.null(date1)) {
    warning("Date mode only provides precision to the day. Use time_lapse mode if more precision is needed.")
    if (anyNA(as.Date(c(date1, date2)))) {
      stop("Error: Dates must be in '%Y-%m-%d' format")
    }
  }

  # end of general arg checks~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  # convert time units to seconds (half_life and time_lapse)~~~~~

  # if time_lapse is given, convert to seconds
  if (!is.null(time_lapse)) {
    time_lapse_sec <- dplyr::case_when(
      time_unit == "year" |
        time_unit == "years" |
        time_unit == "y" ~
      time_lapse * 3.15576e7,

      time_unit == "days" |
        time_unit == "d" |
        time_unit == "day" ~
      time_lapse * 86400,

      time_unit == "hours" |
        time_unit == "h" |
        time_unit == "hour" ~
      time_lapse * 3600,

      time_unit == "minutes" |
        time_unit == "m" |
        time_unit == "mins" ~
      time_lapse * 60,

      time_unit == "seconds" |
        time_unit == "secs" |
        time_unit == "s" ~
      time_lapse
    )
  }

  # end of time_lapse unit conversion


  # get decay constant (in decays / s), RN_select option
  if (!is.null(RN_select)) {
    i <- which(RadData::ICRP_07.NDX$RN == RN_select)
    if (i == 0) stop(paste0("data entered for radionuclide: ", RN_select, " not found."))
    decay_constant <- as.numeric(RadData::ICRP_07.NDX[i, 32])
  }

  if (!is.null(RN_select)) {
    print.data.frame(RadData::ICRP_07.NDX[i, 1:4], row.names = "")
    cat("\n")
  }

  # convert decay_constant (when needed)
  if (!is.null(half_life)) {
    decay_constant <- dplyr::case_when(
      time_unit == "year" |
        time_unit == "years" |
        time_unit == "y" ~
      log(2) / (half_life * 3.15576e7),

      time_unit == "days" |
        time_unit == "d" |
        time_unit == "day" ~
      log(2) / (half_life * 86400),

      time_unit == "hours" |
        time_unit == "h" |
        time_unit == "hour" ~
      log(2) / (half_life * 3600),

      time_unit == "minutes" |
        time_unit == "m" |
        time_unit == "mins" ~
      log(2) / (half_life * 60),

      time_unit == "seconds" |
        time_unit == "secs" |
        time_unit == "s" ~
      log(2) / (half_life)
    )
  }


  # compute time_lapse in seconds from dates (when needed)
  if (!is.null(date1)) {
    time_lapse_sec <- as.numeric(difftime(strptime(date2, "%Y-%m-%d"),
      strptime(date1, "%Y-%m-%d"),
      units = "secs"
    ))
  }

  # Time-decayed response, A2

  A2 <- A1 * exp(-decay_constant * time_lapse_sec)

  # Four cases:
  #   RN_select and date1
  #   RN_select and time_lapse
  #   half_life and date1
  #   half_life and time_lapse
  # Each of these will use the decay formula:

  if (!is.null(RN_select) & !is.null(date1)) {
    n <- length(date2)
    df <- data.frame(
      "RN" = rep(RN_select, n),
      "RefValue" = rep(A1, n),
      "RefDate" = rep(date1, n),
      "TargDate" = date2,
      "dk_value" = A2
    )
  }

  if (!is.null(RN_select) & !is.null(time_lapse)) {
    n <- length(time_lapse)
    df <- data.frame(
      "RN" = rep(RN_select, n),
      "RefValue" = rep(A1, n),
      "DecayTime" = time_lapse,
      "TimeUnit" = rep(time_unit, n),
      "dk_value" = A2
    )
  }

  if ((!is.null(half_life) & !is.null(date1))) {
    n <- length(date2)
    df <- data.frame(
      "half_life" = rep(half_life, n),
      "RefValue" = rep(A1, n),
      "RefDate" = rep(date1, n),
      "TargDate" = date2,
      "dk_value" = A2
    )
  }

  if (!is.null(half_life) & !is.null(time_lapse)) {
    n <- length(time_lapse)
    df <- data.frame(
      "half_life" = rep(half_life, n),
      "RefValue" = rep(A1, n),
      "DecayTime" = time_lapse,
      "TimeUnit" = rep(time_unit, n),
      "dk_value" = A2
    )
  }

  if (num != TRUE)  {
    print.data.frame(df, row.names = FALSE)
    }
  if (num == TRUE) {
    return(A2)
  }
}
